### ISSUE TITLE
How to write custom loss functions to train global models on multivariate time-series

### ISSUE THREAD

**Issue by andreakiro** [unit8co/darts#1374]
I would like to run a bunch of experiments and compare performances of some of the global models (e.g. `NBeats`) your library provides, on my multiple multivariate time-series data. I've read some of your doc and in [this tutorial notebook](https://github.com/unit8co/darts/blob/master/examples/01-multi-time-series-and-covariates.ipynb) you mention that global models can seamlessly move from univariate to multivariate time-series, which sounds cool. But after quickly going through the codebase, it looks like the architecture considers only (multi)variables solving a same objective, in this case regressions.

To be more explicit on my use-case, my (multiple) time-series each have three distinct variables. One is a continuous number to be predicted and is a very traditional case of time-series forecasting, but the other two are categorical variables, let's say discrete labels from 1 to 10. From what I've read the training of your [TorchForecastingModel](https://github.com/unit8co/darts/blob/master/darts/models/forecasting/torch_forecasting_model.py) are built on top of `pylightning` trainers, and from my understanding you don't specify an objective function, do you?

I could already make use of your models and architecture off-the-shell as it is, considering all three variables as regression variables. I would still need to post-process the outputs to clamp categorical ones in their respective classes, and this is manageable. But what bothers me the most is that I would be training these two variables with regression metrics and cannot make use of more accurate classification-based metrics that I would need to fine-tune my results.

So, I wanted to know if I missed something that could allow to train according to a custom objective function?

**Comment by madtoinou** [comment]
Hi!

You can actually pass an loss function to the `Pytorch-Lightning`trainer using the keyword `loss_fn` when you declare your model. By default, the module is using the `torch.nn.MSELoss()`. 

One way to use a different objective functions for each component of your multivariate TimeSeries would be to define a custom loss function that would take care of slicing the target and prediction tensor before applying different objective to each component. I am not sure if this approach would yield accurate predictions since the weights of the model will be shared across the components and you would still have to perform some kind of post-processing for the categorical variables (eventually using a wrapper around the `predict` method to call softmax on a subset of the columns for example). 

Another approach would be to split your multivariate `TimeSeries` into three univariate ones and to train a model with the appropriate objective function for each, while using the other two as covariate... I don't know what makes the most sense with your data.

**Comment by andreakiro** [comment]
Hi @madtoinou. Sorry for my late reply I missed your answer! Thank you very much for the help and good insights. Using three univariate models using covariates to train every objective function independently make sense (and is easier as well). Only "concern" with this approach vs. the global model is that I was hoping for the global model to capture some deeper patterns on my dataset by training them simultaneously, as my 3 target variables are not independent and probably influence each other (but this is my prior intuition, not base knowledge).
